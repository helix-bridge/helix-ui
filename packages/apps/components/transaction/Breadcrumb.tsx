import { Breadcrumb } from 'antd';
import { useTranslation } from 'next-i18next';

export function IBreadcrumb({ txHash }: { txHash?: string }) {
  const { t } = useTranslation();

  return (
    <Breadcrumb
      className="whitespace-nowrap flex items-center overflow-hidden overflow-ellipsis"
      style={{ marginTop: '1.5rem' }}
    >
      <Breadcrumb.Item>{t('Explorer')}</Breadcrumb.Item>
      <Breadcrumb.Item>{t('Transaction')}</Breadcrumb.Item>
      <Breadcrumb.Item>
        <span className="w-32 md:w-72 lg:w-96 truncate">{txHash}</span>
      </Breadcrumb.Item>
    </Breadcrumb>
  );
}
