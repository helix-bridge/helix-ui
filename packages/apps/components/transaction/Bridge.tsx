import { useRouter } from 'next/router';
import { useMemo } from 'react';
import { Logo } from 'shared/components/widget/Logo';
import { ChainConfig, Network } from 'shared/model';
import { getBridge } from '../../utils/bridge';
import { getChainConfig, getDisplayName } from '../../utils/network';

interface BridgeProps {
  from?: ChainConfig;
  to?: ChainConfig;
  size?: 'small' | 'default';
}

export function Bridge({ from, to, size = 'default' }: BridgeProps) {
  const router = useRouter();
  const measure = { default: 40, small: 20 }[size];

  const [departure, arrival] = useMemo(
    () => [from ?? getChainConfig(router.query.from as Network), to ?? getChainConfig(router.query.to as Network)],
    [from, router.query.from, router.query.to, to]
  );

  const bridge = getBridge([departure, arrival]);

  return (
    <div>
      <div
        className="flex justify-between items-center gap-4 bg-gray-200 dark:bg-antDark bg-opacity-25"
        style={{ borderRadius: measure }}
      >
        <div className="p-2 pr-0 flex items-center">
          <Logo chain={departure} width={measure} height={measure} className="w-5 md:w-10" />
        </div>

        <div
          className="flex items-center justify-center self-stretch px-4 md:px-8 bg-gray-300 bg-opacity-20 dark:bg-blue-900 dark:bg-opacity-25 relative"
          style={{
            clipPath: 'polygon(85% 0%, 100% 50%, 85% 100%, 0% 100%, 15% 50%, 0% 0%)',
          }}
        >
          {/* eslint-disable-next-line @next/next/no-img-element */}
          <img
            src={`/image/bridges/${bridge.category}.png`}
            className={size === 'small' ? 'w-5 md:w-14' : `w-10 md:w-28`}
          />
        </div>

        <div className="p-2 pl-0 flex items-center">
          <Logo chain={arrival} width={measure} height={measure} className="w-5 md:w-10" />
        </div>
      </div>

      <div className="flex justify-between text-xs capitalize mt-1 px-3 whitespace-nowrap">
        {[departure, arrival].map((item) => (
          <div key={item.name}>{getDisplayName(item)}</div>
        ))}
      </div>
    </div>
  );
}
