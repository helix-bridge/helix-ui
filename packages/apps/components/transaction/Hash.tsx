import { ExplorerLink } from 'shared/components/widget/ExplorerLink';
import { Network } from 'shared/model';

interface HashProps {
  hash: string;
  network: Network;
}

export const Hash = ({ hash, network }: HashProps) => {
  const txHash = hash.includes('-') ? undefined : hash;
  const extrinsic = hash.includes('-') ? hash.split('-') : undefined;

  return (
    <ExplorerLink
      network={network}
      txHash={txHash}
      copyable
      extrinsic={extrinsic && { height: extrinsic[0], index: extrinsic[1] }}
      className="hover:opacity-80 transition-opacity duration-200 underline text-white"
    ></ExplorerLink>
  );
};
