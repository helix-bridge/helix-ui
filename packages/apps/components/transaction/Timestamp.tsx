import { ClockCircleOutlined } from '@ant-design/icons';
import { Divider, Progress } from 'antd';
import { format, formatDistance, fromUnixTime } from 'date-fns';
import { useTranslation } from 'next-i18next';
import { Icon } from 'shared/components/widget/Icon';
import { DATE_TIME_FORMAT } from 'shared/config/constant';
import { HelixHistoryRecord } from 'shared/model';
import { TransferDescription } from './TransferDescription';

export function Timestamp({ record }: { record: HelixHistoryRecord | null }) {
  const { t } = useTranslation();

  return (
    <TransferDescription
      title={t('Timestamp')}
      tip={t('Date & time of cross-chain transaction inclusion, including length of time for confirmation.')}
    >
      {record && (
        <div className="flex items-center gap-2 whitespace-nowrap">
          {record.result ? <ClockCircleOutlined /> : <Icon name="reload" className="w-4 h-4" />}

          {formatDistance(fromUnixTime(record.startTime), new Date(new Date().toUTCString()), {
            includeSeconds: true,
            addSuffix: true,
          })}

          <span className="hidden md:inline-block">({format(fromUnixTime(record.startTime), DATE_TIME_FORMAT)})</span>

          <Divider type="vertical" orientation="center" />

          <Icon name="clock-fill" className="text-gray-400 text-base w-4 h-4" />

          {record.startTime && record.endTime ? (
            <span className="text-gray-400">
              {t('Confirmed within {{des}}', {
                des: formatDistance(fromUnixTime(record.endTime), fromUnixTime(record.startTime), {
                  includeSeconds: true,
                }),
              })}
            </span>
          ) : (
            <div className="w-32">
              <Progress
                strokeColor={{
                  from: '#108ee9',
                  to: '#87d068',
                }}
                percent={99.9}
                status="active"
                showInfo={false}
              />
            </div>
          )}
        </div>
      )}
    </TransferDescription>
  );
}
