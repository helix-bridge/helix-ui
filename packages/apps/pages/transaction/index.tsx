import { GetServerSidePropsContext } from 'next';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';
import dynamic from 'next/dynamic';
import History from '../../components/history/History';
import { AccountProvider, ApiProvider, ClaimProvider, TxProvider, WalletProvider } from '../../providers';

const ActiveAccount = dynamic(() => import('../../components/widget/account/ActiveAccount'), { ssr: false });

function Page() {
  return (
    <ApiProvider>
      <WalletProvider>
        <AccountProvider>
          <TxProvider>
            <ClaimProvider>
              <div id="app-header-container" className="hidden lg:flex items-center space-x-4 fixed top-4 z-50">
                <ActiveAccount />
              </div>

              <History />
            </ClaimProvider>
          </TxProvider>
        </AccountProvider>
      </WalletProvider>
    </ApiProvider>
  );
}

export const getServerSideProps = async ({ locale, res }: GetServerSidePropsContext) => {
  const translations = await serverSideTranslations(locale ?? 'en', ['common']);
  res.setHeader('Cache-Control', 'public, s-maxage=60, stale-while-revalidate=100');

  return {
    props: {
      ...translations,
    },
  };
};

export default Page;
