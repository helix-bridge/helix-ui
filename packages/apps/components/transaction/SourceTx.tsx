import { useTranslation } from 'next-i18next';
import { useRouter } from 'next/router';
import { Network } from 'shared/model';
import { Hash } from './Hash';
import { TransferDescription } from './TransferDescription';

export function SourceTx({ hash }: { hash: string | undefined }) {
  const { t } = useTranslation();
  const router = useRouter();

  return (
    <TransferDescription
      title={t('Source Tx Hash')}
      tip={t('Unique character string (TxID) assigned to every verified transaction on the Source Chain.')}
    >
      {hash && <Hash network={router.query.from as Network} hash={hash} />}
    </TransferDescription>
  );
}
