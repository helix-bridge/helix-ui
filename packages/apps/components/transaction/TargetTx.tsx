import { useTranslation } from 'next-i18next';
import { useMemo } from 'react';
import { RecordStatus } from 'shared/config/constant';
import { HelixHistoryRecord } from 'shared/model';
import { isCBridgeRecord, isHelixRecord, isXCMRecord } from '../../utils/record/record';
import { Hash } from './Hash';
import { TransferDescription } from './TransferDescription';

export function TargetTx({ record }: { record: HelixHistoryRecord | null }) {
  const { t } = useTranslation();

  // eslint-disable-next-line complexity
  const content = useMemo(() => {
    if (!record) {
      return null;
    }

    if ([RecordStatus.pending, RecordStatus.pendingToClaim, RecordStatus.pendingToRefund].includes(record.result)) {
      return <span>{t('Waiting for the transaction on the target chain...')}</span>;
    }

    const departure = record.fromChain;
    const arrival = record.toChain;

    if (isHelixRecord(record)) {
      if (record.result === RecordStatus.success) {
        return <Hash network={arrival} hash={record.responseTxHash} />;
      }

      if (record.result === RecordStatus.refunded) {
        return record.responseTxHash && <Hash network={departure} hash={record.responseTxHash} />;
      }
    } else if (isCBridgeRecord(record)) {
      return (
        <Hash network={record.result === RecordStatus.refunded ? departure : arrival} hash={record.responseTxHash} />
      );
    } else if (isXCMRecord(record)) {
      return <Hash network={arrival} hash={record.responseTxHash} />;
    }

    return null;
  }, [record, t]);

  return (
    <TransferDescription
      title={t('Target Tx Hash')}
      tip={t('Unique character string (TxID) assigned to every verified transaction on the Target Chain.')}
    >
      {content}
    </TransferDescription>
  );
}
