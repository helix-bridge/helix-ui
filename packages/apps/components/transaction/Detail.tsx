import { Divider } from 'antd';
import ErrorBoundary from 'antd/lib/alert/ErrorBoundary';
import { useMemo } from 'react';
import { TextWithCopy } from 'shared/components/widget/TextWithCopy';
import { useITranslation } from 'shared/hooks';
import { BridgeCategory, HelixHistoryRecord } from 'shared/model';
import { revertAccount } from 'shared/utils/helper/address';
import { fromWei, prettyNumber } from 'shared/utils/helper/balance';
import { isCrabDVMHeco } from 'utils/bridge';
import { getChainConfig } from 'utils/network';
import { TransferStep } from '../../model/transfer';
import { ClaimProvider, TxProvider } from '../../providers';
import { IBreadcrumb } from './Breadcrumb';
import { Bridge } from './Bridge';
import { SourceTx } from './SourceTx';
import { TargetTx } from './TargetTx';
import { Timestamp } from './Timestamp';
import { Token } from './Token';
import { TransferDescription } from './TransferDescription';
import { TransferDetail } from './TransferDetail';
import { TxStatus } from './TxStatus';

interface DetailProps {
  record: HelixHistoryRecord;
  transfers: TransferStep[];
}

export function Detail({ record, transfers }: DetailProps) {
  const { t } = useITranslation();
  const departure = getChainConfig(record.fromChain);
  const arrival = getChainConfig(record.toChain);

  const amount = useMemo(() => {
    const token = departure.tokens.find((item) => item.symbol.toLowerCase() === record.sendToken.toLowerCase());

    return fromWei({ value: record.sendAmount ?? 0, decimals: token?.decimals ?? 9 }, prettyNumber);
  }, [departure.tokens, record]);

  const feeDecimals = useMemo(() => {
    const token = departure.tokens.find((item) => item.symbol.toLowerCase() === record.feeToken.toLowerCase());

    return token?.decimals ?? 9;
  }, [departure.tokens, record]);

  return (
    <TxProvider>
      <ClaimProvider>
        <IBreadcrumb txHash={record.requestTxHash} />

        <h3 className="uppercase text-xs md:text-lg font-bold my-6">{t('transaction detail')}</h3>

        <div className="px-8 py-3 bg-antDark">
          <ErrorBoundary>
            <Bridge from={departure} to={arrival} category={record.bridge.split('-')[0] as BridgeCategory} />
          </ErrorBoundary>

          <Divider />

          <ErrorBoundary>
            <TxStatus record={record} />
          </ErrorBoundary>

          <ErrorBoundary>
            <SourceTx record={record} />
          </ErrorBoundary>

          <ErrorBoundary>
            <TargetTx record={record} />
          </ErrorBoundary>

          <ErrorBoundary>
            <Timestamp record={record} />
          </ErrorBoundary>

          <Divider />

          <TransferDescription title={t('Sender')} tip={t('Address (external or contract) sending the transaction.')}>
            <TextWithCopy>{revertAccount(record.sender, departure)}</TextWithCopy>
          </TransferDescription>

          <TransferDescription
            title={t('Receiver')}
            tip={t('Address (external or contract) receiving the transaction.')}
          >
            <TextWithCopy>{revertAccount(record.recipient, arrival)}</TextWithCopy>
          </TransferDescription>

          {!!transfers.length && <TransferDetail transfers={transfers} />}

          <ErrorBoundary>
            <Token record={record} />
          </ErrorBoundary>

          <Divider />

          <TransferDescription
            title={t('Value')}
            tip={t('The amount to be transferred to the recipient with the cross-chain transaction.')}
          >
            {amount} {transfers[0]?.token?.name}
          </TransferDescription>

          <TransferDescription
            title={t('Transaction Fee')}
            tip={'Amount paid for processing the cross-chain transaction.'}
          >
            {fromWei({ value: record.fee, decimals: feeDecimals })}{' '}
            {record.feeToken === 'null' ? null : record.feeToken}
          </TransferDescription>

          <Divider />

          <TransferDescription title={t('Nonce')} tip={t('A unique number of cross-chain transaction in Bridge')}>
            {isCrabDVMHeco(record.fromChain, record.toChain) ? record.messageNonce : record.nonce}
          </TransferDescription>
        </div>
      </ClaimProvider>
    </TxProvider>
  );
}
