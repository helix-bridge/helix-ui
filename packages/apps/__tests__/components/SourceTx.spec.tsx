/// <reference types="jest" />

import { create } from 'react-test-renderer';
import { SourceTx } from '../../components/transaction/SourceTx';

jest.mock('next-i18next', () => ({
  useTranslation() {
    return {
      t: (label: string) => label,
    };
  },
}));

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: {},
      asPath: '',
    };
  },
}));

describe('<SourceTx />', () => {
  it('render SourceTx', () => {
    const component = create(<SourceTx hash="0xa9673370dce7f48b306fb96b90c42fa2a13986f1e618eb2996eaeb5abce88643" />);

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
