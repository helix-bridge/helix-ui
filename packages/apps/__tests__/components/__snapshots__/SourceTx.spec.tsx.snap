// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SourceTx /> render SourceTx 1`] = `
<div
  className="flex flex-wrap md:items-center gap-2 md:gap-16 my-3 md:my-6 relative"
>
  <div
    className="w-36 flex items-center gap-2"
  >
    <span
      aria-label="info-circle"
      className="anticon anticon-info-circle cursor-help"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="info-circle"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
        />
        <path
          d="M464 336a48 48 0 1096 0 48 48 0 10-96 0zm72 112h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V456c0-4.4-3.6-8-8-8z"
        />
      </svg>
    </span>
    <span
      className="capitalize"
    >
      Source Tx Hash
    </span>
  </div>
  <span
    className="space-x-2"
  >
    <a
      className="hover:opacity-80 transition-opacity duration-200 underline text-white"
      href="https://undefined.etherscan.io/tx/0xa9673370dce7f48b306fb96b90c42fa2a13986f1e618eb2996eaeb5abce88643"
      onClick={[Function]}
      rel="noreferrer"
      target="_blank"
    >
      0xa9673370dce7f48b306fb96b90c42fa2a13986f1e618eb2996eaeb5abce88643
    </a>
    <svg
      className="icon text-base w-4 h-4 cursor-pointer "
      onClick={[Function]}
    />
  </span>
</div>
`;
